File: TradingSwarm/src/pipeline/oms.py
"""
Order Management System — parent/child order model.

Responsibilities:
- Accept only risk-approved order intents
- Maintain canonical order state machine
- Translate parent intents into child venue orders
- Process fills/cancels/rejects
- Detect orphaned or stale working orders
- Survive restart and resume safely

All state transitions are DB-first + outbox.
"""

from __future__ import annotations

import logging
import uuid
from datetime import datetime, timezone
from decimal import Decimal
from typing import Any

from src.core.event_schema import (
    Fill,
    OrderEvent,
    OrderIntent,
    OrderIntentState,
    StreamEnvelope,
    VenueOrder,
    VenueOrderState,
)
from src.ledger import outbox, postgres

logger = logging.getLogger(__name__)

# Valid parent intent state transitions
INTENT_TRANSITIONS: dict[OrderIntentState, set[OrderIntentState]] = {
    OrderIntentState.INTENT_CREATED: {OrderIntentState.RISK_PENDING},
    OrderIntentState.RISK_PENDING: {
        OrderIntentState.RISK_APPROVED,
        OrderIntentState.REJECTED,
    },
    OrderIntentState.RISK_APPROVED: {OrderIntentState.RESERVING_BUDGET},
    OrderIntentState.RESERVING_BUDGET: {
        OrderIntentState.ACCEPTED_BY_OMS,
        OrderIntentState.REJECTED,
    },
    OrderIntentState.ACCEPTED_BY_OMS: {OrderIntentState.ROUTING},
    OrderIntentState.ROUTING: {
        OrderIntentState.WORKING,
        OrderIntentState.REJECTED,
        OrderIntentState.CANCELED,
    },
    OrderIntentState.WORKING: {
        OrderIntentState.PARTIALLY_FILLED,
        OrderIntentState.FILLED,
        OrderIntentState.CANCELED,
        OrderIntentState.SUSPENDED,
    },
    OrderIntentState.PARTIALLY_FILLED: {
        OrderIntentState.PARTIALLY_FILLED,
        OrderIntentState.FILLED,
        OrderIntentState.CANCELED,
        OrderIntentState.SUSPENDED,
    },
    OrderIntentState.SUSPENDED: {
        OrderIntentState.WORKING,
        OrderIntentState.CANCELED,
    },
}

# Valid child venue order state transitions
VENUE_ORDER_TRANSITIONS: dict[VenueOrderState, set[VenueOrderState]] = {
    VenueOrderState.CHILD_CREATED: {VenueOrderState.SUBMITTED},
    VenueOrderState.SUBMITTED: {
        VenueOrderState.ACKNOWLEDGED,
        VenueOrderState.REJECTED,
        VenueOrderState.UNKNOWN_BUT_OPEN,
    },
    VenueOrderState.ACKNOWLEDGED: {
        VenueOrderState.PARTIALLY_FILLED,
        VenueOrderState.FILLED,
        VenueOrderState.CANCEL_REQUESTED,
        VenueOrderState.CANCELED,
        VenueOrderState.EXPIRED,
    },
    VenueOrderState.PARTIALLY_FILLED: {
        VenueOrderState.PARTIALLY_FILLED,
        VenueOrderState.FILLED,
        VenueOrderState.CANCEL_REQUESTED,
        VenueOrderState.CANCELED,
    },
    VenueOrderState.CANCEL_REQUESTED: {
        VenueOrderState.CANCELED,
        VenueOrderState.FILLED,
        VenueOrderState.PARTIALLY_FILLED,
    },
    VenueOrderState.UNKNOWN_BUT_OPEN: {
        VenueOrderState.ACKNOWLEDGED,
        VenueOrderState.FILLED,
        VenueOrderState.CANCELED,
        VenueOrderState.REJECTED,
    },
}


class OrderManagementSystem:
    """
    Parent/child OMS with DB-first state persistence.

    All mutations go through:
    1. Validate state transition
    2. Write to Postgres (+ outbox) in a transaction
    3. Publish via outbox worker
    """

    def __init__(self) -> None:
        self._frozen = True  # frozen until reconciliation passes

    @property
    def is_frozen(self) -> bool:
        return self._frozen

    def unfreeze(self) -> None:
        logger.info("OMS unfrozen — trading enabled")
        self._frozen = False

    def freeze(self, reason: str = "") -> None:
        logger.warning("OMS frozen: %s", reason or "manual freeze")
        self._frozen = True

    # ---- Parent intent operations ----

    async def accept_intent(self, intent: OrderIntent) -> OrderIntent:
        """
        Accept a risk-approved order intent into the OMS.
        Deduplicates on idempotency_key.
        """
        if self._frozen:
            raise RuntimeError("OMS is frozen — no new orders until reconciliation passes")

        if intent.current_state != OrderIntentState.RISK_APPROVED:
            raise ValueError(
                f"OMS only accepts risk_approved intents, got {intent.current_state}"
            )

        pool = await postgres.get_pool()
        async with pool.acquire() as conn:
            # Dedupe check
            existing = await conn.fetchrow(
                "SELECT order_intent_id, current_state FROM order_intents WHERE idempotency_key = $1",
                intent.idempotency_key,
            )
            if existing:
                logger.info("Duplicate intent: %s", intent.idempotency_key)
                intent.order_intent_id = existing["order_intent_id"]
                intent.current_state = OrderIntentState(existing["current_state"])
                return intent

            intent.current_state = OrderIntentState.ACCEPTED_BY_OMS
            intent.updated_at = datetime.now(timezone.utc)

            async with conn.transaction():
                await conn.execute(
                    """
                    INSERT INTO order_intents (
                        order_intent_id, idempotency_key, team_id, strategy_id,
                        instrument_id, venue_preference, side, intent_type,
                        requested_qty, limit_price, stop_price, time_in_force,
                        signal_id, correlation_id, model_version, feature_version,
                        config_hash, current_state
                    ) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16,$17,$18)
                    """,
                    intent.order_intent_id, intent.idempotency_key,
                    intent.team_id, intent.strategy_id,
                    intent.instrument_id, intent.venue_preference,
                    intent.side.value, intent.intent_type.value,
                    intent.requested_qty, intent.limit_price,
                    intent.stop_price, intent.time_in_force.value,
                    intent.signal_id, intent.correlation_id,
                    intent.model_version, intent.feature_version,
                    intent.config_hash, intent.current_state.value,
                )

                envelope = StreamEnvelope.wrap(
                    "order_intent.accepted",
                    intent,
                    idempotency_key=intent.idempotency_key,
                )
                await outbox.write_with_outbox(
                    conn,
                    "SELECT 1",  # no additional business SQL needed
                    (),
                    "oms.intents",
                    envelope,
                )

        logger.info(
            "Intent accepted: %s [%s %s %s]",
            intent.order_intent_id, intent.side.value,
            intent.requested_qty, intent.instrument_id,
        )
        return intent

    async def transition_intent(
        self,
        order_intent_id: uuid.UUID,
        new_state: OrderIntentState,
        event_type: str = "",
        payload: dict[str, Any] | None = None,
    ) -> None:
        """Transition a parent intent to a new state with validation."""
        pool = await postgres.get_pool()
        async with pool.acquire() as conn:
            row = await conn.fetchrow(
                "SELECT current_state FROM order_intents WHERE order_intent_id = $1 FOR UPDATE",
                order_intent_id,
            )
            if row is None:
                raise ValueError(f"Unknown order intent: {order_intent_id}")

            current = OrderIntentState(row["current_state"])
            allowed = INTENT_TRANSITIONS.get(current, set())
            if new_state not in allowed:
                raise ValueError(
                    f"Invalid transition: {current} -> {new_state} "
                    f"(allowed: {allowed})"
                )

            now = datetime.now(timezone.utc)
            async with conn.transaction():
                await conn.execute(
                    "UPDATE order_intents SET current_state = $1, updated_at = $2 WHERE order_intent_id = $3",
                    new_state.value, now, order_intent_id,
                )

                event = OrderEvent(
                    order_intent_id=order_intent_id,
                    event_type=event_type or f"intent.{new_state.value}",
                    event_time_utc=now,
                    payload=payload or {},
                )
                await conn.execute(
                    """
                    INSERT INTO order_events (event_id, order_intent_id, event_type, event_time_utc, payload)
                    VALUES ($1, $2, $3, $4, $5)
                    """,
                    event.event_id, order_intent_id,
                    event.event_type, event.event_time_utc,
                    event.payload,
                )

        logger.info("Intent %s: %s -> %s", order_intent_id, current.value, new_state.value)

    # ---- Child venue order operations ----

    async def create_child_order(
        self,
        order_intent_id: uuid.UUID,
        venue: str,
        requested_qty: Decimal,
        limit_price: Decimal | None = None,
    ) -> VenueOrder:
        """Create a child venue order for a parent intent."""
        pool = await postgres.get_pool()
        async with pool.acquire() as conn:
            # Get next child sequence
            max_seq = await conn.fetchval(
                "SELECT COALESCE(MAX(child_seq), 0) FROM venue_orders WHERE order_intent_id = $1",
                order_intent_id,
            )
            child_seq = (max_seq or 0) + 1

            child = VenueOrder(
                order_intent_id=order_intent_id,
                venue=venue,
                child_seq=child_seq,
                requested_qty=requested_qty,
                remaining_qty=requested_qty,
                limit_price=limit_price,
            )

            await conn.execute(
                """
                INSERT INTO venue_orders (
                    venue_order_id_internal, order_intent_id, venue, child_seq,
                    current_state, requested_qty, remaining_qty, limit_price
                ) VALUES ($1,$2,$3,$4,$5,$6,$7,$8)
                """,
                child.venue_order_id_internal, order_intent_id,
                venue, child_seq,
                child.current_state.value, requested_qty,
                requested_qty, limit_price,
            )

        logger.info(
            "Child order created: %s (parent=%s, seq=%d, venue=%s)",
            child.venue_order_id_internal, order_intent_id, child_seq, venue,
        )
        return child

    async def transition_venue_order(
        self,
        venue_order_id: uuid.UUID,
        new_state: VenueOrderState,
        broker_order_id: str | None = None,
        filled_qty: Decimal | None = None,
        avg_fill_price: Decimal | None = None,
    ) -> None:
        """Transition a child venue order with validation."""
        pool = await postgres.get_pool()
        async with pool.acquire() as conn:
            row = await conn.fetchrow(
                "SELECT current_state, order_intent_id FROM venue_orders WHERE venue_order_id_internal = $1 FOR UPDATE",
                venue_order_id,
            )
            if row is None:
                raise ValueError(f"Unknown venue order: {venue_order_id}")

            current = VenueOrderState(row["current_state"])
            allowed = VENUE_ORDER_TRANSITIONS.get(current, set())
            if new_state not in allowed:
                raise ValueError(
                    f"Invalid venue order transition: {current} -> {new_state}"
                )

            now = datetime.now(timezone.utc)
            updates = ["current_state = $1", "updated_at = $2"]
            args: list[Any] = [new_state.value, now]
            idx = 3

            if broker_order_id is not None:
                updates.append(f"broker_order_id = ${idx}")
                args.append(broker_order_id)
                idx += 1
            if filled_qty is not None:
                updates.append(f"filled_qty = ${idx}")
                args.append(filled_qty)
                idx += 1
                updates.append(f"remaining_qty = requested_qty - ${idx}")
                args.append(filled_qty)
                idx += 1
            if avg_fill_price is not None:
                updates.append(f"avg_fill_price = ${idx}")
                args.append(avg_fill_price)
                idx += 1

            args.append(venue_order_id)
            set_clause = ", ".join(updates)
            await conn.execute(
                f"UPDATE venue_orders SET {set_clause} WHERE venue_order_id_internal = ${idx}",
                *args,
            )

            # Log event
            await conn.execute(
                """
                INSERT INTO order_events (
                    event_id, order_intent_id, venue_order_id_internal,
                    event_type, broker_order_id, event_time_utc
                ) VALUES ($1, $2, $3, $4, $5, $6)
                """,
                uuid.uuid4(), row["order_intent_id"], venue_order_id,
                f"venue_order.{new_state.value}", broker_order_id, now,
            )

        logger.info("Venue order %s: %s -> %s", venue_order_id, current.value, new_state.value)

    # ---- Fill processing ----

    async def record_fill(self, fill: Fill) -> None:
        """Record a fill and update parent/child state."""
        pool = await postgres.get_pool()
        async with pool.acquire() as conn:
            async with conn.transaction():
                # Insert fill
                await conn.execute(
                    """
                    INSERT INTO fills (
                        fill_id, order_intent_id, venue_order_id_internal,
                        instrument_id, team_id, strategy_id, venue, side,
                        quantity, price, fee, fee_currency,
                        expected_fill_price, slippage_bps, fill_time_utc
                    ) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15)
                    """,
                    fill.fill_id, fill.order_intent_id,
                    fill.venue_order_id_internal, fill.instrument_id,
                    fill.team_id, fill.strategy_id, fill.venue,
                    fill.side.value, fill.quantity, fill.price,
                    fill.fee, fill.fee_currency,
                    fill.expected_fill_price, fill.slippage_bps,
                    fill.fill_time_utc,
                )

                # Update venue order filled qty
                if fill.venue_order_id_internal:
                    await conn.execute(
                        """
                        UPDATE venue_orders SET
                            filled_qty = filled_qty + $1,
                            remaining_qty = requested_qty - filled_qty - $1,
                            avg_fill_price = $2,
                            updated_at = NOW()
                        WHERE venue_order_id_internal = $3
                        """,
                        fill.quantity, fill.price,
                        fill.venue_order_id_internal,
                    )

                # Outbox for downstream
                envelope = StreamEnvelope.wrap(
                    "fill.recorded", fill,
                    idempotency_key=str(fill.fill_id),
                )
                await outbox.write_with_outbox(
                    conn,
                    "SELECT 1", (),
                    "fills.events", envelope,
                )

        logger.info(
            "Fill recorded: %s [%s %s @ %s on %s]",
            fill.fill_id, fill.side.value, fill.quantity, fill.price, fill.venue,
        )

    # ---- Restart recovery (Section 10.5) ----

    async def load_unfinished(self) -> tuple[list[dict], list[dict]]:
        """Load all unfinished intents and venue orders on restart."""
        pool = await postgres.get_pool()
        async with pool.acquire() as conn:
            intents = await conn.fetch("""
                SELECT * FROM order_intents
                WHERE current_state NOT IN ('filled', 'canceled', 'rejected', 'expired')
                ORDER BY created_at
            """)
            venue_orders = await conn.fetch("""
                SELECT * FROM venue_orders
                WHERE current_state NOT IN ('filled', 'canceled', 'rejected', 'expired')
                ORDER BY updated_at
            """)

        logger.info(
            "Loaded %d unfinished intents, %d unfinished venue orders",
            len(intents), len(venue_orders),
        )
        return [dict(r) for r in intents], [dict(r) for r in venue_orders]

    async def detect_stale_orders(self, stale_threshold_seconds: int = 300) -> list[dict]:
        """Find orders that have been working too long without update."""
        pool = await postgres.get_pool()
        async with pool.acquire() as conn:
            rows = await conn.fetch("""
                SELECT * FROM venue_orders
                WHERE current_state IN ('submitted', 'acknowledged', 'unknown_but_open')
                AND updated_at < NOW() - $1 * INTERVAL '1 second'
            """, stale_threshold_seconds)
        if rows:
            logger.warning("Found %d stale venue orders", len(rows))
        return [dict(r) for r in rows]

File: Julia_research/QuantEngine/src/production/oms.jl
# ── Order Management System — Parent/Child with State Machine ──────────
# DB-first: all state transitions written to Postgres before anything else.
# No order without a durable intent. No fill without a durable record.
#
# Parent intent = strategic desire ("buy $10K BTC")
# Child venue order = tactical execution ("submit 0.2 BTC @ market on Binance")

using Dates
using UUIDs

# ── State Machines ────────────────────────────────────────────

@enum IntentState begin
    INTENT_CREATED
    RISK_PENDING
    RISK_APPROVED
    RESERVING_BUDGET
    ACCEPTED_BY_OMS
    ROUTING
    WORKING
    PARTIALLY_FILLED
    FILLED
    CANCELED
    REJECTED
    EXPIRED
    SUSPENDED
end

@enum VenueOrderState begin
    CHILD_CREATED
    SUBMITTED
    ACKNOWLEDGED
    VENUE_PARTIALLY_FILLED
    VENUE_FILLED
    CANCEL_REQUESTED
    VENUE_CANCELED
    VENUE_REJECTED
    VENUE_EXPIRED
    UNKNOWN_BUT_OPEN
end

# Valid parent intent transitions — DAG, no backward edges
const INTENT_TRANSITIONS = Dict{IntentState, Set{IntentState}}(
    INTENT_CREATED     => Set([RISK_PENDING]),
    RISK_PENDING       => Set([RISK_APPROVED, REJECTED]),
    RISK_APPROVED      => Set([RESERVING_BUDGET]),
    RESERVING_BUDGET   => Set([ACCEPTED_BY_OMS, REJECTED]),
    ACCEPTED_BY_OMS    => Set([ROUTING]),
    ROUTING            => Set([WORKING, REJECTED, CANCELED]),
    WORKING            => Set([PARTIALLY_FILLED, FILLED, CANCELED, SUSPENDED]),
    PARTIALLY_FILLED   => Set([PARTIALLY_FILLED, FILLED, CANCELED, SUSPENDED]),
    SUSPENDED          => Set([WORKING, CANCELED]),
)

# Terminal states — no further transitions
const TERMINAL_INTENT_STATES = Set([FILLED, CANCELED, REJECTED, EXPIRED])

# Valid child venue order transitions
const VENUE_TRANSITIONS = Dict{VenueOrderState, Set{VenueOrderState}}(
    CHILD_CREATED         => Set([SUBMITTED]),
    SUBMITTED             => Set([ACKNOWLEDGED, VENUE_REJECTED, UNKNOWN_BUT_OPEN]),
    ACKNOWLEDGED          => Set([VENUE_PARTIALLY_FILLED, VENUE_FILLED, CANCEL_REQUESTED, VENUE_CANCELED, VENUE_EXPIRED]),
    VENUE_PARTIALLY_FILLED => Set([VENUE_PARTIALLY_FILLED, VENUE_FILLED, CANCEL_REQUESTED, VENUE_CANCELED]),
    CANCEL_REQUESTED      => Set([VENUE_CANCELED, VENUE_FILLED, VENUE_PARTIALLY_FILLED]),
    UNKNOWN_BUT_OPEN      => Set([ACKNOWLEDGED, VENUE_FILLED, VENUE_CANCELED, VENUE_REJECTED]),
)

"""Check if a state transition is valid."""
function valid_transition(from::IntentState, to::IntentState)::Bool
    allowed = get(INTENT_TRANSITIONS, from, Set{IntentState}())
    return to in allowed
end

function valid_venue_transition(from::VenueOrderState, to::VenueOrderState)::Bool
    allowed = get(VENUE_TRANSITIONS, from, Set{VenueOrderState}())
    return to in allowed
end

# ── Data Types ────────────────────────────────────────────────

struct OrderIntent
    order_intent_id::UUID
    idempotency_key::String
    team_id::String
    strategy_id::String
    instrument_id::String      # canonical instrument identifier
    venue_preference::String
    side::Symbol               # :buy, :sell
    intent_type::Symbol        # :market, :limit, :stop, :stop_limit
    requested_qty::Float64
    limit_price::Float64       # 0.0 if market
    stop_price::Float64        # 0.0 if not stop
    time_in_force::Symbol      # :gtc, :ioc, :fok, :day
    signal_id::String
    correlation_id::UUID
    model_version::String
    feature_version::String
    config_hash::String
    current_state::IntentState
    created_at::DateTime
end

struct VenueOrder
    venue_order_id::UUID
    order_intent_id::UUID
    venue::String
    child_seq::Int
    broker_order_id::String
    current_state::VenueOrderState
    requested_qty::Float64
    submitted_qty::Float64
    filled_qty::Float64
    remaining_qty::Float64
    limit_price::Float64
    avg_fill_price::Float64
    submitted_at::Union{DateTime, Nothing}
end

struct Fill
    fill_id::UUID
    order_intent_id::UUID
    venue_order_id::UUID
    instrument_id::String
    team_id::String
    strategy_id::String
    venue::String
    side::Symbol
    quantity::Float64
    price::Float64
    fee::Float64
    fee_currency::String
    expected_price::Float64
    slippage_bps::Float64
    fill_time::DateTime
end

# ── OMS Core ──────────────────────────────────────────────────

mutable struct OrderManagementSystem
    pool::Any                  # PgPool — typed as Any to avoid circular deps
    frozen::Bool
    lock::ReentrantLock
end

function OrderManagementSystem(pool)
    return OrderManagementSystem(pool, true, ReentrantLock())  # starts FROZEN
end

"""Accept a risk-approved intent into the OMS. Deduplicates on idempotency_key."""
function accept_intent!(oms::OrderManagementSystem, intent::OrderIntent)
    if oms.frozen
        error("OMS is frozen — no new orders until reconciliation passes")
    end

    if intent.current_state != RISK_APPROVED
        error("OMS only accepts RISK_APPROVED intents, got $(intent.current_state)")
    end

    with_transaction(oms.pool) do conn
        # Dedupe check
        existing = LibPQ.execute(conn,
            "SELECT order_intent_id, current_state FROM order_intents WHERE idempotency_key = \$1",
            [intent.idempotency_key]
        )
        existing_data = columntable(existing)
        close(existing)

        if !isempty(first(values(existing_data)))
            @info "Duplicate intent: $(intent.idempotency_key)"
            return nothing
        end

        # Insert intent as ACCEPTED_BY_OMS
        LibPQ.execute(conn, """
            INSERT INTO order_intents (
                order_intent_id, idempotency_key, team_id, strategy_id,
                instrument_id, venue_preference, side, intent_type,
                requested_qty, limit_price, stop_price, time_in_force,
                signal_id, correlation_id, model_version, feature_version,
                config_hash, current_state
            ) VALUES (\$1,\$2,\$3,\$4,\$5,\$6,\$7,\$8,\$9,\$10,\$11,\$12,\$13,\$14,\$15,\$16,\$17,\$18)
        """, [
            string(intent.order_intent_id), intent.idempotency_key,
            intent.team_id, intent.strategy_id,
            intent.instrument_id, intent.venue_preference,
            string(intent.side), string(intent.intent_type),
            intent.requested_qty, intent.limit_price,
            intent.stop_price, string(intent.time_in_force),
            intent.signal_id, string(intent.correlation_id),
            intent.model_version, intent.feature_version,
            intent.config_hash, "ACCEPTED_BY_OMS",
        ])

        # Write order event
        LibPQ.execute(conn, """
            INSERT INTO order_events (event_id, order_intent_id, event_type, event_time_utc)
            VALUES (\$1, \$2, 'intent.accepted_by_oms', NOW())
        """, [string(uuid4()), string(intent.order_intent_id)])
    end

    @info "Intent accepted: $(intent.order_intent_id) [$(intent.side) $(intent.requested_qty)]"
    return intent
end

"""Transition a parent intent to a new state. Validates the state machine."""
function transition_intent!(oms::OrderManagementSystem, intent_id::UUID, new_state::IntentState;
                            event_type::String="", payload::String="{}")
    with_transaction(oms.pool) do conn
        # Lock the row
        result = LibPQ.execute(conn,
            "SELECT current_state FROM order_intents WHERE order_intent_id = \$1 FOR UPDATE",
            [string(intent_id)]
        )
        data = columntable(result)
        close(result)

        if isempty(first(values(data)))
            error("Unknown intent: $intent_id")
        end

        current_str = first(values(data))[1]
        current = parse_intent_state(current_str)

        if !valid_transition(current, new_state)
            error("Invalid transition: $current → $new_state")
        end

        LibPQ.execute(conn, """
            UPDATE order_intents SET current_state = \$1, updated_at = NOW()
            WHERE order_intent_id = \$2
        """, [string(new_state), string(intent_id)])

        etype = isempty(event_type) ? "intent.$(lowercase(string(new_state)))" : event_type
        LibPQ.execute(conn, """
            INSERT INTO order_events (event_id, order_intent_id, event_type, event_time_utc, payload)
            VALUES (\$1, \$2, \$3, NOW(), \$4)
        """, [string(uuid4()), string(intent_id), etype, payload])
    end

    @info "Intent $intent_id: → $new_state"
end

"""Create a child venue order for a parent intent."""
function create_child_order!(oms::OrderManagementSystem, intent_id::UUID, venue::String,
                             qty::Float64; limit_price::Float64=0.0)
    child_id = uuid4()
    child_seq = 1

    with_connection(oms.pool) do conn
        # Get next sequence
        result = LibPQ.execute(conn,
            "SELECT COALESCE(MAX(child_seq), 0) FROM venue_orders WHERE order_intent_id = \$1",
            [string(intent_id)]
        )
        data = columntable(result)
        close(result)
        max_seq = first(values(data))[1]
        child_seq = (max_seq === nothing ? 0 : max_seq) + 1

        LibPQ.execute(conn, """
            INSERT INTO venue_orders (
                venue_order_id_internal, order_intent_id, venue, child_seq,
                current_state, requested_qty, remaining_qty, limit_price
            ) VALUES (\$1,\$2,\$3,\$4,\$5,\$6,\$7,\$8)
        """, [
            string(child_id), string(intent_id), venue, child_seq,
            "CHILD_CREATED", qty, qty, limit_price,
        ])
    end

    @info "Child order: $child_id (parent=$intent_id, seq=$child_seq, venue=$venue)"
    return (id=child_id, seq=child_seq)
end

"""Record a fill and update order states."""
function record_fill!(oms::OrderManagementSystem, fill::Fill)
    with_transaction(oms.pool) do conn
        # Insert fill
        LibPQ.execute(conn, """
            INSERT INTO fills (
                fill_id, order_intent_id, venue_order_id_internal,
                instrument_id, team_id, strategy_id, venue, side,
                quantity, price, fee, fee_currency,
                expected_fill_price, slippage_bps, fill_time_utc
            ) VALUES (\$1,\$2,\$3,\$4,\$5,\$6,\$7,\$8,\$9,\$10,\$11,\$12,\$13,\$14,\$15)
        """, [
            string(fill.fill_id), string(fill.order_intent_id),
            string(fill.venue_order_id), fill.instrument_id,
            fill.team_id, fill.strategy_id, fill.venue,
            string(fill.side), fill.quantity, fill.price,
            fill.fee, fill.fee_currency,
            fill.expected_price, fill.slippage_bps,
            string(fill.fill_time),
        ])

        # Update venue order
        LibPQ.execute(conn, """
            UPDATE venue_orders SET
                filled_qty = filled_qty + \$1,
                remaining_qty = requested_qty - filled_qty - \$1,
                avg_fill_price = \$2,
                updated_at = NOW()
            WHERE venue_order_id_internal = \$3
        """, [fill.quantity, fill.price, string(fill.venue_order_id)])
    end

    @info "Fill: $(fill.fill_id) [$(fill.side) $(fill.quantity) @ $(fill.price)]"
end

"""Load all unfinished intents and venue orders (for restart recovery)."""
function load_unfinished(oms::OrderManagementSystem)
    intents = pg_fetch(oms.pool, """
        SELECT * FROM order_intents
        WHERE current_state NOT IN ('FILLED', 'CANCELED', 'REJECTED', 'EXPIRED')
        ORDER BY created_at
    """)

    orders = pg_fetch(oms.pool, """
        SELECT * FROM venue_orders
        WHERE current_state NOT IN ('VENUE_FILLED', 'VENUE_CANCELED', 'VENUE_REJECTED', 'VENUE_EXPIRED')
        ORDER BY updated_at
    """)

    @info "Loaded $(length(first(values(intents)))) unfinished intents, $(length(first(values(orders)))) unfinished venue orders"
    return (intents=intents, orders=orders)
end

# ── State Parsing ─────────────────────────────────────────────

function parse_intent_state(s::AbstractString)::IntentState
    mapping = Dict(
        "INTENT_CREATED" => INTENT_CREATED,
        "RISK_PENDING" => RISK_PENDING,
        "RISK_APPROVED" => RISK_APPROVED,
        "RESERVING_BUDGET" => RESERVING_BUDGET,
        "ACCEPTED_BY_OMS" => ACCEPTED_BY_OMS,
        "ROUTING" => ROUTING,
        "WORKING" => WORKING,
        "PARTIALLY_FILLED" => PARTIALLY_FILLED,
        "FILLED" => FILLED,
        "CANCELED" => CANCELED,
        "REJECTED" => REJECTED,
        "EXPIRED" => EXPIRED,
        "SUSPENDED" => SUSPENDED,
    )
    return get(mapping, uppercase(s), INTENT_CREATED)
end
